props/websubmit/src/main.rs

```diff
@@ -85,16 +93,98 @@ impl DeletionProp {
     }
 }
 
+/// Storing data in the database must be associated to a user. This is
+/// necessary for e.g. the deletion to work.
+pub struct ScopedStorageProp {
+    cx: Arc<PolicyContext>,
+}
+
+impl ScopedStorageProp {
+    pub fn new(cx: Arc<PolicyContext>) -> Self {
+        ScopedStorageProp { cx }
+    }
+
+    pub fn check(self) {
+        for c_id in self.cx.desc().controllers.keys() {
+            let scopes = self
+                .cx
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|node| self.cx.has_marker(marker!(scopes), *node))
+                .collect::<Vec<_>>();
+            let stores = self
+                .cx
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|node| self.cx.has_marker(marker!(stores), *node))
+                .collect::<Vec<_>>();
+            let mut sensitives = self
+                .cx
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|node| self.cx.has_marker(marker!(sensitive), *node));
+
+            let controller_valid = sensitives.all(|sens| {
+                stores.iter().all(|&store| {
+                    // sensitive flows to store implies some scope flows to store callsite
+                    !(self
+                        .cx
+                        .flows_to(sens, store, paralegal_policy::EdgeType::Data))
+                        || store.typ.as_call_site().is_some_and(|cs| {
```

Comment:
Why are we converting to a call site here instead of just using the node?

Reply:
We want to get the CallSite attached to the Node, if it is a Callargument. `scope` can flow to any CallArgument on that same CallSite, not just the one that `store` flows to.

Reply:
Ah I see. I hadn't considered that. Maybe a good idea would be to have a function like `Node::associated_call_site()` that returns `Option<Node>`, trying to convert it to a call site. That would make it more obvious what's happening. Also presumably we'll need this functionality more often in the future.

Reply:
Makes sense, done